### PR TITLE
feat: multi instance support

### PR DESCRIPTION
**Pull request by TobiaszCudnik**
Because

- user can have more than 1 instance of Instill AI

This commit

- adds instance management and switching

## Changes
- instances management
  - `add`, `edit`, `list`, `remove`, `set-default`
- markdown tables
- tests
- typed config (for hosts)
- default config when empty
- hostname validation
- breaking dependabot updates
- readme

### CODE REVIEW

File: pkg/cmd/instances/edit_test.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+	"bytes"
+	"github.com/instill-ai/cli/pkg/iostreams"
+	"testing"
+
+	"github.com/google/shlex"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/instill-ai/cli/internal/config"
+	"github.com/instill-ai/cli/pkg/cmdutil"
+)
```
Comment (pinglin) [inline pkg/cmd/instances/edit_test.go:13]:
```suggestion
import (
	"bytes"
	"testing"

	"github.com/google/shlex"
	"github.com/stretchr/testify/assert"

	"github.com/instill-ai/cli/internal/config"
	"github.com/instill-ai/cli/pkg/cmdutil"
	"github.com/instill-ai/cli/pkg/iostreams"
)
```

File: pkg/cmd/instances/instances.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
+package instances
+
+import (
+	"github.com/instill-ai/cli/internal/config"
+	"github.com/spf13/cobra"
+
+	"github.com/instill-ai/cli/pkg/cmdutil"
+)
```
Comment (pinglin) [inline pkg/cmd/instances/instances.go:8]:
```suggestion
import (
	"github.com/spf13/cobra"

	"github.com/instill-ai/cli/internal/config"
	"github.com/instill-ai/cli/pkg/cmdutil"
)
```

File: pkg/cmd/instances/list.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
+package instances
+
+import (
+	"fmt"
+	"github.com/MakeNowJust/heredoc"
+	"github.com/instill-ai/cli/internal/config"
+	"github.com/instill-ai/cli/pkg/cmdutil"
+	"github.com/instill-ai/cli/pkg/iostreams"
+	"github.com/spf13/cobra"
+)
```
Comment (pinglin) [inline pkg/cmd/instances/list.go:10]:
```suggestion
import (
	"fmt"
	
	"github.com/spf13/cobra"
	
	"github.com/MakeNowJust/heredoc"
	"github.com/instill-ai/cli/internal/config"
	"github.com/instill-ai/cli/pkg/cmdutil"
	"github.com/instill-ai/cli/pkg/iostreams"
)
```

File: pkg/cmd/instances/list_test.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+	"bytes"
+	"github.com/instill-ai/cli/pkg/iostreams"
+	"testing"
+
+	"github.com/google/shlex"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/instill-ai/cli/internal/config"
+	"github.com/instill-ai/cli/pkg/cmdutil"
+)
```
Comment (pinglin) [inline pkg/cmd/instances/list_test.go:13]:
```suggestion
import (
	"bytes"
	"testing"

	"github.com/google/shlex"
	"github.com/stretchr/testify/assert"

	"github.com/instill-ai/cli/internal/config"
	"github.com/instill-ai/cli/pkg/cmdutil"
	"github.com/instill-ai/cli/pkg/iostreams"
)
```

File: pkg/cmd/instances/remove.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+import (
+	"fmt"
+	"github.com/MakeNowJust/heredoc"
+	"github.com/instill-ai/cli/internal/config"
+	"github.com/instill-ai/cli/internal/instance"
+	"github.com/instill-ai/cli/pkg/cmdutil"
+	"github.com/instill-ai/cli/pkg/iostreams"
+	"github.com/spf13/cobra"
+)
```
Comment (pinglin) [inline pkg/cmd/instances/remove.go:11]:
```suggestion
import (
	"fmt"
	
	"github.com/MakeNowJust/heredoc"
	"github.com/spf13/cobra"

	"github.com/instill-ai/cli/internal/config"
	"github.com/instill-ai/cli/internal/instance"
	"github.com/instill-ai/cli/pkg/cmdutil"
	"github.com/instill-ai/cli/pkg/iostreams"
)
```

File: pkg/cmd/instances/remove_test.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+	"bytes"
+	"github.com/instill-ai/cli/pkg/iostreams"
+	"testing"
+
+	"github.com/google/shlex"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/instill-ai/cli/internal/config"
+	"github.com/instill-ai/cli/pkg/cmdutil"
+)
```
Comment (pinglin) [inline pkg/cmd/instances/remove_test.go:13]:
```suggestion
import (
	"bytes"
	"testing"

	"github.com/google/shlex"
	"github.com/stretchr/testify/assert"

	"github.com/instill-ai/cli/internal/config"
	"github.com/instill-ai/cli/pkg/cmdutil"
	"github.com/instill-ai/cli/pkg/iostreams"
)
```

File: pkg/cmd/instances/set_default.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+import (
+	"fmt"
+	"github.com/MakeNowJust/heredoc"
+	"github.com/instill-ai/cli/internal/config"
+	"github.com/instill-ai/cli/internal/instance"
+	"github.com/instill-ai/cli/pkg/cmdutil"
+	"github.com/instill-ai/cli/pkg/iostreams"
+	"github.com/spf13/cobra"
+)
```
Comment (pinglin) [inline pkg/cmd/instances/set_default.go:11]:
```suggestion
import (
	"fmt"

	"github.com/MakeNowJust/heredoc"
	"github.com/spf13/cobra"

	"github.com/instill-ai/cli/internal/config"
	"github.com/instill-ai/cli/internal/instance"
	"github.com/instill-ai/cli/pkg/cmdutil"
	"github.com/instill-ai/cli/pkg/iostreams"
)
```

File: pkg/cmd/instances/set_default_test.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+	"bytes"
+	"github.com/instill-ai/cli/pkg/iostreams"
+	"testing"
+
+	"github.com/google/shlex"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/instill-ai/cli/internal/config"
+	"github.com/instill-ai/cli/pkg/cmdutil"
+)
```
Comment (pinglin) [inline pkg/cmd/instances/set_default_test.go:13]:
```suggestion
import (
	"bytes"
	"testing"

	"github.com/google/shlex"
	"github.com/stretchr/testify/assert"

	"github.com/instill-ai/cli/internal/config"
	"github.com/instill-ai/cli/pkg/cmdutil"
	"github.com/instill-ai/cli/pkg/iostreams"
)
```

File: pkg/cmdutil/auth_check_test.go [hunk 9]
```diff
@@ -1,6 +1,8 @@
 package cmdutil
 
 import (
+	"os"
+	"path/filepath"
```
Comment (pinglin) [inline pkg/cmdutil/auth_check_test.go:5]: please re-arrange import packages

File: internal/oauth2/auth_code_flow.go [hunk 10]
```diff
@@ -83,26 +112,21 @@ func (a *Authenticator) VerifyIDToken(ctx context.Context, token *oauth2.Token)
...
-	auth, err := NewAuthenticator(issuer, clientID, clientSecret, callbackHost, port)
+	issuer := host.APIHostname
+	if host.Oauth2Issuer != "" {
+		issuer = host.Oauth2Issuer
+	}
+	audience := host.APIHostname
+	if host.Oauth2Audience != "" {
+		audience = host.Oauth2Audience
+	}
+	auth, err := NewAuthenticator(issuer, host.Oauth2ClientID, host.Oauth2Secret, callbackHost, port)
```
Comment (pinglin) [inline internal/oauth2/auth_code_flow.go:129]:
```suggestion
	auth, err := NewAuthenticator(issuer, host.Oauth2ClientID, host.Oauth2ClientSecret, callbackHost, port)
```

File: pkg/cmd/instances/add.go [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+	}
+
+	host := config.DefaultHostConfig()
+	host.APIHostname = opts.APIHostname
+	host.IsDefault = opts.Default
+	host.Oauth2Hostname = opts.Oauth2
+	host.Oauth2Audience = opts.Audience
+	host.Oauth2Issuer = opts.Issuer
+	host.Oauth2ClientID = opts.ClientID
+	host.Oauth2Secret = opts.Secret
```
Comment (pinglin) [inline pkg/cmd/instances/add.go:109]:
```suggestion
	host.Oauth2ClientSecret = opts.ClientSecret
```

File: pkg/cmd/instances/add.go [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+	}
+
+	apiHost := opts.APIHostname
+	for _, h := range hosts {
+		if h.APIHostname == apiHost {
+			return fmt.Errorf("ERROR: instance '%s' already exists", apiHost)
+		}
+	}
+
+	if opts.Oauth2 != "" && (opts.Secret == "" || opts.ClientID == "") {
```
Comment (pinglin) [inline pkg/cmd/instances/add.go:98]:
```suggestion
	if opts.Oauth2 != "" && (opts.ClientSecret == "" || opts.ClientID == "") {
```

File: pkg/cmd/instances/edit.go [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+			break
+		}
+	}
+	if host == nil {
+		return fmt.Errorf("ERROR: instance '%s' does not exists", apiHost)
+	}
+
+	if opts.NoAuth {
+		host.Oauth2Issuer = ""
+		host.Oauth2Secret = ""
```
Comment (pinglin) [inline pkg/cmd/instances/edit.go:104]:
```suggestion
		host.Oauth2ClientSecret = ""
```

File: pkg/cmd/instances/edit.go [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+		host.Oauth2ClientID = ""
+		host.Oauth2Hostname = ""
+		host.Oauth2Audience = ""
+	} else if opts.Oauth2 != "" {
+		if (host.Oauth2Secret == "" || host.Oauth2ClientID == "") && (opts.Secret == "" || opts.ClientID == "") {
```
Comment (pinglin) [inline pkg/cmd/instances/edit.go:109]:
```suggestion
		if (host.Oauth2ClientSecret == "" || host.Oauth2ClientID == "") && (opts.ClientSecret == "" || opts.ClientID == "") {
```

File: pkg/cmd/instances/edit.go [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+		}
+		host.Oauth2Hostname = opts.Oauth2
+		if opts.Audience != "" {
+			host.Oauth2Audience = opts.Audience
+		}
+		if opts.Issuer != "" {
+			host.Oauth2Issuer = opts.Issuer
+		}
+		if opts.Secret != "" {
+			host.Oauth2Secret = opts.Secret
```
Comment (pinglin) [inline pkg/cmd/instances/edit.go:120]:
```suggestion
			host.Oauth2ClientSecret = opts.ClientSecret
```

File: internal/config/from_file.go [hunk 16]
```diff
@@ -248,6 +292,130 @@ func (c *fileConfig) parseHosts(hostsEntry *yaml.Node) ([]*HostConfig, error) {
...
+	// instance info
+	APIHostname string `example:"api.instill.tech"`
+	IsDefault   bool
+	APIVersion  string `example:"v1alpha"`
+
+	// oauth config
+	Oauth2Hostname string `example:"auth.instill.tech"`
+	Oauth2Audience string `example:"https://instill.tech"`
+	Oauth2Issuer   string `example:"https://auth.instill.tech/"`
+	Oauth2Secret   string
```
Comment (pinglin) [inline internal/config/from_file.go:332]:
```suggestion
	Oauth2ClientSecret   string
```

File: internal/config/from_file.go [hunk 17]
```diff
@@ -248,6 +292,130 @@ func (c *fileConfig) parseHosts(hostsEntry *yaml.Node) ([]*HostConfig, error) {
...
+	}
+	ht.TokenType = conf.GetOptionalStringValue("token_type")
+	ht.AccessToken = conf.GetOptionalStringValue("access_token")
+	ht.Expiry = conf.GetOptionalStringValue("expiry")
+	ht.RefreshToken = conf.GetOptionalStringValue("refresh_token")
+	ht.IDToken = conf.GetOptionalStringValue("id_token")
+	ht.Oauth2Audience = conf.GetOptionalStringValue("oauth2_audience")
+	ht.Oauth2Issuer = conf.GetOptionalStringValue("oauth2_issuer")
+	ht.Oauth2Hostname = conf.GetOptionalStringValue("oauth2_hostname")
+	ht.Oauth2Secret = conf.GetOptionalStringValue("oauth2_secret")
```
Comment (pinglin) [inline internal/config/from_file.go:364]:
```suggestion
	ht.Oauth2ClientSecret = conf.GetOptionalStringValue("oauth2_client_secret")
```

File: internal/config/from_file.go [hunk 18]
```diff
@@ -248,6 +292,130 @@ func (c *fileConfig) parseHosts(hostsEntry *yaml.Node) ([]*HostConfig, error) {
...
+	}
+	err = conf.SetStringValue("oauth2_hostname", host.Oauth2Hostname)
+	if err != nil {
+		return err
+	}
+	err = conf.SetStringValue("oauth2_client_id", host.Oauth2ClientID)
+	if err != nil {
+		return err
+	}
+	err = conf.SetStringValue("oauth2_secret", host.Oauth2Secret)
```
Comment (pinglin) [inline internal/config/from_file.go:408]:
```suggestion
	err = conf.SetStringValue("oauth2_client_secret", host.Oauth2ClientSecret)
```

File: internal/config/from_file_test.go [hunk 19]
```diff
@@ -1,15 +1,48 @@
...
+	fc := NewBlankConfig()
+	hc := fc.MakeConfigForHost("foo")
+	_ = hc.SetStringValue("oauth2_hostname", "bar1")
+	_ = hc.SetStringValue("oauth2_secret", "bar2")
+	_ = hc.SetStringValue("api_version", "bar3")
+	hct, err := hostConfigToTyped(hc)
+	assert.NoError(t, err)
+
+	assert.Equal(t, "bar1", hct.Oauth2Hostname)
+	assert.Equal(t, "bar2", hct.Oauth2Secret)
```
Comment (pinglin) [inline internal/config/from_file_test.go:41]:
```suggestion
	assert.Equal(t, "bar2", hct.Oauth2ClientSecret)
```

File: internal/config/from_file_test.go [hunk 20]
```diff
@@ -1,15 +1,48 @@
...
+func Test_fileConfig_Typed(t *testing.T) {
+	configDir := filepath.Join(t.TempDir(), ".config", "instill")
+	_ = os.MkdirAll(configDir, 0755)
+	os.Setenv(INSTILL_CONFIG_DIR, configDir)
+	defer os.Unsetenv(INSTILL_CONFIG_DIR)
+
+	fc := NewBlankConfig()
+	hc := fc.MakeConfigForHost("foo")
+	_ = hc.SetStringValue("oauth2_hostname", "bar1")
+	_ = hc.SetStringValue("oauth2_secret", "bar2")
```
Comment (pinglin) [inline internal/config/from_file_test.go:35]:
```suggestion
	_ = hc.SetStringValue("oauth2_client_secret", "bar2")
```

File: internal/config/stub.go [hunk 21]
```diff
@@ -53,3 +55,52 @@ func (c ConfigStub) DefaultHost() (string, error) {
...
+	_ = c.Set(h, "token_type", host.TokenType)
+	_ = c.Set(h, "access_token", host.AccessToken)
+	_ = c.Set(h, "expiry", host.Expiry)
+	_ = c.Set(h, "refresh_token", host.RefreshToken)
+	_ = c.Set(h, "id_token", host.IDToken)
+	_ = c.Set(h, "oauth2_audience", host.Oauth2Audience)
+	_ = c.Set(h, "oauth2_issuer", host.Oauth2Issuer)
+	_ = c.Set(h, "oauth2_hostname", host.Oauth2Hostname)
+	_ = c.Set(h, "oauth2_client_id", host.Oauth2ClientID)
+	_ = c.Set(h, "oauth2_secret", host.Oauth2Secret)
```
Comment (pinglin) [inline internal/config/stub.go:94]:
```suggestion
	_ = c.Set(h, "oauth2_client_secret", host.Oauth2ClientSecret)
```

File: internal/oauth2/auth_code_flow.go [hunk 22]
```diff
@@ -47,6 +49,33 @@ type Authenticator struct {
...
+func HostConfigInstillCloud() *config.HostConfigTyped {
+	initEnv()
+	host := config.DefaultHostConfig()
+	host.APIHostname = "api.instill.tech"
+	host.IsDefault = true
+	host.Oauth2Hostname = hostname
+	host.Oauth2Audience = audience
+	host.Oauth2Issuer = issuer
+	host.Oauth2ClientID = clientID
+	host.Oauth2Secret = clientSecret
```
Comment (pinglin) [inline internal/oauth2/auth_code_flow.go:75]:
```suggestion
	host.Oauth2ClientSecret = clientSecret
```

File: pkg/cmd/auth/login/login.go [hunk 23]
```diff
@@ -82,12 +72,34 @@ func loginRun(f *cmdutil.Factory, opts *LoginOptions) error {
...
+		if h.APIHostname == hostname {
+			host = &h
+			break
+		}
+	}
+	if host == nil {
+		return fmt.Errorf("ERROR: instance '%s' does not exists", hostname)
+	}
+
+	if host.Oauth2Hostname == "" || host.Oauth2ClientID == "" || host.Oauth2Secret == "" {
```
Comment (pinglin) [inline pkg/cmd/auth/login/login.go:91]:
```suggestion
	if host.Oauth2Hostname == "" || host.Oauth2ClientID == "" || host.Oauth2ClientSecret == "" {
```

File: pkg/cmd/instances/add_test.go [hunk 24]
```diff
@@ -0,0 +1,181 @@
...
+					Oauth2:      "bar1",
+					Secret:      "bar2",
+					ClientID:    "bar3",
+				},
+			},
+			expectFn: func(t *testing.T, cfg config.Config) {
+				v, err := cfg.Get("foo", "oauth2_hostname")
+				assert.NoError(t, err)
+				assert.Equal(t, "bar1", v)
+				v, err = cfg.Get("foo", "oauth2_secret")
```
Comment (pinglin) [inline pkg/cmd/instances/add_test.go:139]:
```suggestion
				v, err = cfg.Get("foo", "oauth2_client_secret")
```

File: pkg/cmd/instances/edit_test.go [hunk 25]
```diff
@@ -0,0 +1,182 @@
...
+					Oauth2:      "bar1",
+					Secret:      "bar2",
+					ClientID:    "bar3",
+				},
+			},
+			expectFn: func(t *testing.T, cfg config.Config) {
+				v, err := cfg.Get("api.instill.tech", "oauth2_hostname")
+				assert.NoError(t, err)
+				assert.Equal(t, "bar1", v)
+				v, err = cfg.Get("api.instill.tech", "oauth2_secret")
```
Comment (pinglin) [inline pkg/cmd/instances/edit_test.go:139]:
```suggestion
				v, err = cfg.Get("api.instill.tech", "oauth2_client_secret")
```

**Review by pinglin — CHANGES_REQUESTED** [review]
Please see the comments and suggested changes.

**Comment by TobiaszCudnik** [comment]
> Please see the comments and suggested changes.

@pinglin Ive implemented the renames and formatted imports. Im happy theres no concerns regarding the code itself. We should incorporate `goimports` into the pipeline.

File: internal/config/stub.go [hunk 26]
```diff
@@ -53,3 +55,52 @@ func (c ConfigStub) DefaultHost() (string, error) {
...
+	ins := []HostConfigTyped{
+		{
+			APIHostname:    "api.instill.tech",
+			IsDefault:      true,
+			APIVersion:     "v1alpha",
+			Oauth2Hostname: "auth.instill.tech",
+			Oauth2Audience: "https://api.instill.tech",
+			Oauth2Issuer:   "https://auth.instill.tech/",
+			Oauth2Secret:   "foobar",
+			Oauth2ClientID: "barfoo",
```
[the remark this answers is not in the document]
Comment (pinglin) [reply internal/config/stub.go]: Given the specialty of this file, we will need to set a skip rule for it in the CodeQL GA workflow to avoid false alarm.

**Comment by pinglin** [comment]
> > Please see the comments and suggested changes.
> 
> @pinglin Ive implemented the renames and formatted imports. Im happy theres no concerns regarding the code itself. We should incorporate `goimports` into the pipeline.

Yeah, that will be very ideal. @praharshjain has proposed and tried it before actually. Any guideline for this?

File: pkg/cmd/auth/logout/logout.go [hunk 27]
```diff
@@ -28,7 +28,7 @@ func NewCmdLogout(f *cmdutil.Factory, runF func(*LogoutOptions) error) *cobra.Co
 	cmd := &cobra.Command{
 		Use:   "logout",
 		Args:  cobra.ExactArgs(0),
-		Short: "Log out of a Instill host",
+		Short: "Log out of an Instill instance",
```
[the remark this answers is not in the document]
Comment (pinglin) [reply pkg/cmd/auth/logout/logout.go:31]: Good point. We make this more explicit then.

```suggestion
		Short: "Log out of an Instill Core or Instill Cloud instance",
```

**Comment by TobiaszCudnik** [comment]
This seems to be the right source:
https://github.com/dnephin/pre-commit-golang

Import order should be linted, like any other required conventions. AFAIK we're using the default golangci-lint config.

File: pkg/cmd/auth/login/login.go [hunk 28]
```diff
@@ -82,12 +72,34 @@ func loginRun(f *cmdutil.Factory, opts *LoginOptions) error {
...
+		e := heredoc.Docf(`ERROR: OAuth2 config isn't complete for '%s'
+
+			You fix it with:
+			$ instill instances edit %s \
+				--oauth2 HOSTNAME \
+				--client-id CLIENT_ID \
+				--secret SECRET`, hostname, hostname)
```
[the remark this answers is not in the document]
Comment (pinglin) [reply pkg/cmd/auth/login/login.go:98]: Prefer explicitly here.

File: internal/config/stub.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (TobiaszCudnik) [reply internal/config/stub.go]: There is no suppression by annotation, only in the UI. It wont pop up unless the line is modified. This is the official way of handling it.

**Review by pinglin — APPROVED** [review]
LGTM

**Pull request merged**